tobler/util/raster_tools.py

```diff
@@ -0,0 +1,63 @@
+"""tools for dealing with rasters."""
+
+import ast
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rasterio as rio
+from rasterio import features
+from shapely.geometry import shape
+
+
+def _parse_geom(geom_str):
+    return shape(ast.literal_eval(geom_str))
+
+
+def extract_raster_features(gdf, raster_path, pixel_types=None, nodata=255):
+    """Generate a geodataframe from raster pixels.
+
+    Parameters
+    ----------
+    gdf : geopandas.GeoDataFrame
+        geodataframe  defining the area of interest. The input raster will be
+        clipped to the extent of the geodataframe
+    raster_path : str
+        path to raster file, such as downloaded from <https://lcviewer.vito.be/download>
+    pixel_types : list-like, optional
+        subset of pixel values to extract, by default None. If None, this function
+        may generate a very large geodataframe
+    nodata : int, optional
+        pixel value denoting "no data" in input raster
+
+    Returns
+    -------
+    geopandas.GeoDataFrame
+        geodataframe whose rows are the zones extracted by the rasterio.features module.
+        The geometry of each zone is the boundary of a contiguous group of pixels with 
+        the same value; the `value` column contains the pixel value of each zone.
+    """
+    with rio.open(raster_path) as src:  # read nationwide file
+
+        raster_crs = src.crs.data
+        gdf = gdf.to_crs(raster_crs)
+        geomask = [gdf.explode().unary_union.__geo_interface__]
```

Comment:
Why are you exploding it first?

Reply:
just got used to blowing stuff up in the h3 func and wasnt sure if rasterio would balk at the multipolys

Reply:
Exploding before `unary_union` will likely make no difference on the result passed to rasterio. At least it should not. But it may be faster/slower to do unary_union.

Reply:
I think actually there might be a benefit in exploding here. I seem to remember if you load a multi poly in rasterio, it'll load its entire extent (this may have changed), which you don't want.

Reply:
The code above flows as `multpolygon -> explode: polygons -> unary_union: multipolygon -> __geo_interface__`. If rasterio has an issue with MultiPolygon the explode above does not help it and we would need to do first unary_union and then loop over its parts below.